### PR TITLE
Sb scraper

### PR DESCRIPTION
**Pull request by popescu2**
Mostly working Spacebattles scraper. Request to merge to master so video can be made showing both scrapers work.

### CODE REVIEW

File: scrapers/Spacebattles/Spacebattles/items.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# -*- coding: utf-8 -*-
+
+# Define here the models for your scraped items
+#
+# See documentation in:
+# http://doc.scrapy.org/en/latest/topics/items.html
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/items.py]: We probably don't need the above comments in this file, do we?

File: scrapers/Spacebattles/Spacebattles/items.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import scrapy
+
+
+class StorySegment(scrapy.Item):
+    # define the fields for your item here like:
+    #  = scrapy.Field()
+    content = scrapy.Field()
+    nsfw_flag = scrapy.Field()
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/items.py]: Are you using this flag?

File: scrapers/Spacebattles/Spacebattles/middlewares.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+
+# Define here the models for your spider middleware
+#
+# See documentation in:
+# http://doc.scrapy.org/en/latest/topics/spider-middleware.html
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/middlewares.py:6]: See previous comment about this comment block (I won't say this on every file, I promise, just this last one).

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+        """
+
+        current_page = response.xpath("//a[@class='currentPage ']/text()")
+        print("current page: {0}".format(current_page.extract_first()))
+
+        next_page_link = response.xpath("//a[@class='text' and contains(., 'Next')]")
+        next_page_link = next_page_link.xpath('@href').extract_first()
+
+        # urls_stories is a touple with a url, and a corresponding Story object
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py]: "tuple," fwiw

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        if next_page_link is not None:
+
+            #print("next page link: {0}".format(next_page_link))
+            next_page_link = response.urljoin(next_page_link)
+            yield scrapy.Request(next_page_link, callback=self.loop_pages, priority=0)
+        """
+
+    def get_thread_urls(self, response):
+
+        """ Scan the threads on a page
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py:69]: Reminding you that you wanted to update this docstring (and maybe do a pass at other docstrings).

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        """
+
+        print("scraping {0}".format(response.url))
+        url_stories = []
+
+        li_tags = response.xpath("//li[@class='discussionListItem visible  ']")
+
+        for thread_tag in li_tags:
+
+            # change this to make sure that an author+host is unique
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py]: Punt this to future work! :)

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+            # print("\nauthors: {0}".format(author.name))
+            # print(  "    url: {0}".format(thread_url))
+            # print(  "  title: {0}".format(story.title))
+
+            if thread_url is not None:
+
+                thread_link = response.urljoin(thread_url)
+                url_stories.append((thread_link, story))
+                # yield scrapy.Request(thread_link, callback=self.scan_thread)
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py]: ?

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    def scan_thread(self, response):
+
+        """ Scan the actual thread for story and author content
+
+        This looks for any threadmarks, processes them, and
+        follows the last threadmark's "next" link.
+        If the page doesn't have any threadmarks, or there is
+        no next link, it gracefully closes. It also creates story
+        segments, and links them to a story. (django models)
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py:121]: A+ specificity here

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                story_seg.title = title
+                story_seg.story = story_item
+
+                date_time = div_tmark.xpath(".//span[@class='DateTime']/@title").extract_first()
+                if date_time is None:
+                    date_time = div_tmark.xpath(".//abbr[@class='DateTime']/text()").extract_first()
+                date_obj = datetime.strptime(date_time, "%b %d, %Y at %I:%M %p")
+                story_seg.published = date_obj
+
+                # TODO ^^ Convert date time to actual date_time type
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py]: Time zones wooo. (Feel free to poke the Tumblr scraper to see how to add them, and/or ask me.)

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+            
+
+
+
+        
+
+
+
+
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py]: Roomy! But probably unnecessary whitespace. ;)

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                # TODO ^^ Convert date time to actual date_time type
+
+                content = div_tmark.xpath(".//blockquote/text()").extract_first().encode('utf-8')
+                content = " ".join(content.split())
+                story_seg.contents = content
+
+                print("Title: {0}   Author: {1}".format(story_seg.title, author))
+                print("date_time: {0}".format(date_obj))
+                # print("content: {0}".format(content[0:200]))
```
Comment (relsqui) [inline scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py:165]: I suspect these are suppose to either both be commented or both be uncommented.

File: scrapers/Spacebattles/readme.md [hunk 12]
```diff
@@ -0,0 +1,17 @@
+Navigate to the "noveltorpedo/scrapers/Spacebattles" directory
+before proceeding.
+
+Install python dependencies:
+```
+pip install -r requirements.txt
+```
+
+Start up elasticsearch and database. (instructions in the website readme)
```
Comment (relsqui) [inline scrapers/Spacebattles/readme.md]: You can put an actual (relative) link here; see the main page readme for an example.

**Review by relsqui — CHANGES_REQUESTED** [review]
GJ! Will work on updating the Tumblr bits myself and touch base again later this week.

File: scrapers/Spacebattles/Spacebattles/items.py (hunk 2 above)
Comment (popescu2) [reply scrapers/Spacebattles/Spacebattles/items.py]: Ah, that code is deprecated.

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py (hunk 7 above)
Comment (popescu2) [reply scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py]: Old code. Removed in latest commit

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py (hunk 9 above)
Comment (popescu2) [reply scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py]: Also fixed!

File: scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py (hunk 11 above)
Comment (popescu2) [reply scrapers/Spacebattles/Spacebattles/spiders/spacebattles_scraper.py:165]: For debugging, it's not always necessary to print out so much text from a story. :/

**Pull request merged**